Develop (Групповой проект Filmorate)

AlexeiRomanchenko opened the pull request.

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Разработать функциональность "Отзывы"
V1

ARGalimov committed:
Merge pull request #14 from AlexeiRomanchenko/add-reviews

add-reviews

ARGalimov committed:
Разработать функциональность "Лента событий"
V1

ARGalimov committed:
Разработать функциональность "Лента событий"
V1

ARGalimov committed:
Разработать функциональность "Лента событий"
V1

ARGalimov committed:
Разработать функциональность "Лента событий"
V1

ARGalimov committed:
Разработать функциональность "Лента событий"
V1

ARGalimov committed:
Разработать функциональность "Лента событий"
V1

ARGalimov committed:
Разработать функциональность "Лента событий"
V1

ARGalimov committed:
Разработать функциональность "Лента событий"
V1

ARGalimov committed:
Разработать функциональность "Лента событий"
V1

ARGalimov committed:
Разработать функциональность "Лента событий"
V1

ARGalimov committed:
Разработать функциональность "Лента событий"
V1

AlexeiRomanchenko committed:
feat: Добавлена Функциональность «Фильмы по режиссёрам».

AlexeiRomanchenko committed:
feat: Добавлена Функциональность «Фильмы по режиссёрам».

AlexeiRomanchenko committed:
feat: Добавлена Функциональность «Фильмы по режиссёрам».

AlexeiRomanchenko committed:
feat: Добавлена Функциональность «Фильмы по режиссёрам».

AlexeiRomanchenko committed:
feat: Добавлена Функциональность «Фильмы по режиссёрам».

AlexeiRomanchenko committed:
feat: Добавлена Функциональность «Фильмы по режиссёрам».

ARGalimov committed:
Разработать функциональность "Лента событий"
V1

elGordoGato committed:
feat: search

AlexeiRomanchenko committed:
feat: Добавлена Функциональность «Общие фильмы».

AlexeiRomanchenko committed:
feat: Добавлена Функциональность «Общие фильмы».

ARGalimov committed:
Merge pull request #20 from AlexeiRomanchenko/add-feed

Разработать функциональность "Лента событий"

elGordoGato committed:
Функциональность «Рекомендации» (#17)

feat: add findRecommendations method

anton-tt committed:
Удаление пользователей и фильмов - попытка №2

AlexeiRomanchenko committed:
fix: исправил замечания по внутреннему ревью

AlexeiRomanchenko committed:
Добавил изменения в схему БД

AlexeiRomanchenko committed:
Update README.md

AlexeiRomanchenko committed:
Merge branch 'develop' of https://github.com/AlexeiRomanchenko/java-filmorate into add-director

# Conflicts:
#	db_diagram.png
#	src/main/java/ru/yandex/practicum/filmorate/model/Film.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/interfaces/FilmStorage.java
#	src/main/resources/application.properties
#	src/main/resources/schema.sql
#	src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java

AlexeiRomanchenko committed:
Сделал pull с develop

AlexeiRomanchenko committed:
Merge remote-tracking branch 'origin/add-director' into add-director

AlexeiRomanchenko committed:
fix:codeStyle

AlexeiRomanchenko committed:
fix:clearTests

AlexeiRomanchenko committed:
Merge branch 'develop' of https://github.com/AlexeiRomanchenko/java-filmorate into add-common-films

# Conflicts:
#	pom.xml
#	src/main/java/ru/yandex/practicum/filmorate/model/Film.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/interfaces/FilmStorage.java
#	src/main/resources/application-film.yml
#	src/main/resources/application.properties
#	src/main/resources/schema.sql
#	src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java

AlexeiRomanchenko committed:
Сделал pull c develop

AlexeiRomanchenko committed:
fix: добавил путь на @NotNull

AlexeiRomanchenko committed:
Merge pull request #27 from AlexeiRomanchenko/add-common-films

anton-tt committed:
Merge pull request #28 from AlexeiRomanchenko/add-remove-endpoint

Удаление пользователей и фильмов - попытка №2

AlexeiRomanchenko committed:
fix: восстановлена add-director после конфликта

AlexeiRomanchenko committed:
fix: восстановлена add-director после конфликта

AlexeiRomanchenko committed:
Merge remote-tracking branch 'origin/add-director' into add-director

AlexeiRomanchenko committed:
Merge branch 'develop' into add-director

AlexeiRomanchenko committed:
fix: исправлен метод delele()[не верный мердж]

AlexeiRomanchenko committed:
Update FilmDbStorage.java

AlexeiRomanchenko committed:
Update FilmDbStorage.java

AlexeiRomanchenko committed:
Добавлена Функциональность «Фильмы по режиссёрам».

Добавлена Функциональность «Фильмы по режиссёрам».

elGordoGato committed:
# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java
#	src/main/java/ru/yandex/practicum/filmorate/description/LogMessagesFilms.java
#	src/main/java/ru/yandex/practicum/filmorate/model/Film.java
#	src/main/java/ru/yandex/practicum/filmorate/service/EventService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/db/DirectorDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/db/EventDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/interfaces/DirecorStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/interfaces/FilmStorage.java
#	src/main/resources/application-director.yml
#	src/main/resources/application-film.yml
#	src/main/resources/schema.sql
#	src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java

elGordoGato committed:
Merge branch 'develop' of https://github

elGordoGato committed:
Merge branch 'develop' of https://github.com/AlexeiRomanchenko/java-filmorate into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java
#	src/main/java/ru/yandex/practicum/filmorate/description/LogMessagesFilms.java
#	src/main/java/ru/yandex/practicum/filmorate/model/Film.java
#	src/main/java/ru/yandex/practicum/filmorate/service/EventService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/db/DirectorDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/db/EventDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/db/UserDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/interfaces/DirecorStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/interfaces/FilmStorage.java
#	src/main/resources/application-director.yml
#	src/main/resources/application-film.yml
#	src/main/resources/schema.sql
#	src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java

elGordoGato committed:
Merge remote-tracking branch 'origin/add-search' into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/db/DirectorDbStorage.java

anton-tt committed:
Список фильмов по популярности с сортировкой по году и жанрам - попытка №2

anton-tt committed:
Исправление недочётов-1

anton-tt committed:
Merge pull request #30 from AlexeiRomanchenko/add-most-populars

Add most populars

elGordoGato committed:
Merge branch 'develop' into add-search

elGordoGato committed:
feat: search (#29)

feat: search

anton-tt committed:
Исправление недочётов

anton-tt committed:
Merge pull request #32 from AlexeiRomanchenko/add-most-populars

Исправление недочётов

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/DirectorController.java:
```diff
@@ -0,0 +1,60 @@
+    public final DirectorService directorService;
+
+    @Autowired
+    public DirectorController(DirectorService directorService) {
```
Если хотите, можно заменить данный конструктор аннотацией `@RequiredArgsConstructor`, от lombok, тогда можно будет просто навесить данную аннотацию над классом и убрать данный код для конструктора

AlexeiRomanchenko replied:
исправил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/DirectorController.java:
```diff
@@ -0,0 +1,60 @@
+
+    @PutMapping("/directors")
+    public Director updateDirector(@Valid @RequestBody Director director) {
+        log.info(LogDirector.PUT_DIRECTOR_BY_ID_REQUEST.getMessage() + director.getId());
```
Рекомендую передавать параметры при логирование через `placeholder - {}`
Можно записать логирование следующим образом:
`log.info("my log {}", myParameter);`

AlexeiRomanchenko replied:
исправил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -63,9 +73,31 @@ public void deleteLike(@PathVariable int id, @PathVariable int userId) {
-    public Collection<Film> getPopularFilms(@RequestParam(value = "count", defaultValue = "10") int count) {
-        log.info(LogMessagesFilms.GET_LIST_POPULAR_FILMS_REQUEST.getMessage());
-        return filmService.getPopularFilms(count);
+    public Collection<Film> getSortedPopularFilms(@RequestParam(defaultValue = "10", value = "count") int count,
```
Если хотите, можно записать аннотации валидаций для проверки ограничений для данных параметров, например здесь можно использовать аннотацию `@Positive`, она не допустит передачи отрицательных значений для данного параметра, аналогично можно сделать и для других подобных мест в коде

AlexeiRomanchenko replied:
исправил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/description/EventType.java:
```diff
@@ -0,0 +1,15 @@
+    FRIEND("FRIEND");
+    private final String event;
+
+    EventType(String event) {
```
Если хотите, данный конструктор можно заменить аннотацией от lombok - `@RequiredArgsConstructor`

ARGalimov replied:
Исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -20,7 +23,18 @@ public class Film {
     private Set<Genre> genres;
+    private List<Director> directors;
     private RatingMpa mpa;
+    private Set<Integer> likes;
```
Данное поле более не требуется, так как лайки теперь у нас хранятся только на стороне БД

AlexeiRomanchenko replied:
Исправлено

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/Review.java:
```diff
@@ -0,0 +1,37 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class Review {
+    private final Map<Integer, Boolean> grades = new HashMap<>();
```
Действительно ли данная мапа требуется?

ARGalimov replied:
Исправил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/Review.java:
```diff
@@ -0,0 +1,37 @@
+    private Integer filmId;
+
+    public int getUseful() {
+        return grades.values().stream().mapToInt(positive -> positive ? 1 : -1).sum();
```
Думаю, можно создать поле `useful` типа `int` и хранить в нем требуемое значение

ARGalimov replied:
Исправил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -20,7 +19,10 @@ public class User {
+    @PastOrPresent
+    @NotNull
     private LocalDate birthday;
+    private Set<Integer> likes;
```
Для какого функционала служит данное поле?

AlexeiRomanchenko replied:
исправил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -2,30 +2,49 @@
 public class FilmService {
     private final UserStorage userStorage;
     private final FilmStorage filmStorage;
+    private final EventService eventService;
```
Рекомендую стараться использовать только хранилища на уровне сервисов

ARGalimov replied:
Исправил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -2,30 +2,49 @@
+    }
+
+    private void checkUserId(Integer id) {
+        if (id < 1 || userStorage.getById(id) == null) {
```
Рекомендую наложить ограничение `@Positive` в модели данных, тогда можно будет не проверять `id`

AlexeiRomanchenko replied:
исправил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -55,8 +79,11 @@ public Film findFilm(int id) {
         Film film = filmStorage.getById(filmId);
         if (film != null) {
+
             if (userStorage.getById(userId) != null) {
```
Здесь и в аналогичных местах можно переиспользовать метод `checkUserId(Integer id)`

AlexeiRomanchenko replied:
исправил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,159 @@
+    public List<Review> findAll() {
+        List<Review> reviews = reviewStorage.findAll();
+        reviewStorage.loadAllGrades(reviews);
+        reviews.sort(Comparator.comparing(Review::getUseful).reversed());
```
Рекомендую вынести компаратор в константу

ARGalimov replied:
Исправил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    public Review findById(Integer id) {
+        Review review = reviewStorage.findById(id).orElseThrow(() -> {
```
Вот это хороший приём с использованием `orElseThrow`, рекомендую сделать так во всём коде

elGordoGato replied:
Done

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/DirectorDbStorage.java:
```diff
@@ -0,0 +1,169 @@
+public class DirectorDbStorage implements DirecorStorage {
+
+    private final JdbcTemplate jdbcTemplate;
+    @Value("${director.get-id-by-director}")
```
Очень круто, что получилось брать свойства и `properties` файла, это очень сложный момент, очень рад, что получилось с этим разобраться!

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java:
```diff
@@ -180,44 +269,50 @@ private Film makeFilm(ResultSet rs, int id) throws SQLException {
-        return film;
     }
 
     private Film filmMap(SqlRowSet srs) {
```
Рекомендую все методы для преобразования `ResultSet` (и подобных объектов) вынести в единый класс, который мы назовём, например, `FilmMapper`, соответственно данный класс будет служить для преобразования `ResultSet` в требуемые нам объекты модели данных

elGordoGato replied:
Done

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java:
```diff
@@ -230,4 +325,67 @@ public void clearDbLikes() {
+        }
+        List<Film> films = new LinkedList<>();
+        while (filmIdRow.next()) {
+            films.add(getById(filmIdRow.getInt("film_id")));
```
Здесь стоит отойти от использования запросов в цикле, это плохо, так как порождает N запросов к БД, что сильно её нагружает
Здесь и далее аналогичное замечание по запросам

elGordoGato replied:
Добавлен метод `loadAllDirectors`, вызов его и метода `addGenreForList ` вынесены в `FilmService`

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java:
```diff
@@ -138,20 +204,43 @@ public void addLike(int filmId, int userId) {
+        List<Film> filmList = jdbcTemplate.query(sqlQuery, this::makeFilm);
+        addGenreForList(filmList);
+        return filmList
+                .stream().peek(f -> f.setDirectors(getDirectors(f.getId())))
```
Здесь стоит отойти от использования запросов в цикле, это плохо, так как порождает N запросов к БД, что сильно её нагружает

elGordoGato replied:
Исправлено по аналогии 
> Добавлен метод `loadAllDirectors`, вызов его и метода `addGenreForList ` вынесены в `FilmService`

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/FilmDbStorage.java:
```diff
@@ -230,4 +325,67 @@ public void clearDbLikes() {
+                        "%" + query.toUpperCase() + "%"));
+        searchParams.replaceAll(param -> {
+            switch (param) {
+                case "title":
```
Рекомендую сформировать перечисление под данные возможные типы и далее использовать его в данном операторе `switch`, это ограничит возможный набор значений

elGordoGato replied:
Добавил enum класс `SearchParam`, но вместо switch показалось удачнее использовать стрим и `getColumn`

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/ReviewDbStorage.java:
```diff
@@ -0,0 +1,132 @@
+        String sql = "INSERT INTO REVIEW_USEFUL (REVIEW_ID, USER_ID, GRADE) VALUES(?, ?, ?)";
+        Map<Integer, Boolean> grades = review.getGrades();
+        for (var grade : grades.entrySet()) {
+            jdbcTemplate.update(sql, review.getReviewId(), grade.getKey(), grade.getValue());
```
Надо отойти от использования обновления в цикле, воспользуйтесь пакетным обновлением, как Вы делали ранее в коде

ARGalimov replied:
Исправлено

KirPyt reviewed (commented):
Всех приветствую)

Надо добавить описание реализованных фич в проекте. Можно сделать это в описании к ПР, либо в readme файле.

Проект выполнен очень хорошо, далее напишу некоторые мысли, которые у меня возникли в ходе проверки Вашего проекта:

1. Стоит попробовать практику ревью кода друг друга, это очень полезный навык. В целом можно позадавать друг другу вопросы о том, почему здесь или там использовано то или иное решение и т.д. Так как это Ваша первая командная работа, то стоит попробовать оценить, насколько у Вас хорошо получилось наладить совместную разработку, разбиение реализации фич между собой и навык перекрёстного ревью.

2. Оцените оптимальность sql запросов, в том числе подвыборки данных там, где можно воспользоваться оператором join. Оцените лишние итерации в самом коде (например, поиск пересечений объектов в логике кода). Так же особенно обратите внимание на циклическое обращение к БД, это сильно перегружает базу и является критическим моментом в коде

3. Подумайте о безопасности Вашего приложения в части программиста-клиента. Как бы Вы подстраховали написанный Вами код от неверного его использования или расширения? Или в худшем случае от изменения

4. Обязательно проверьте общий код стайла Вашего проекта - код стайл во всем проекте должен быть выдержан в едином стиле, не содержать лишних переносов, закомментированных участков кода, пробелов и т.д

5. Не забывайте оставлять хорошие комментарии в коде, это поможет всем, и тем кто будет работать с Вашим кодом в дальнейшем, и Вам, когда и если будете писать документацию к своему коду

Если у кого то из Вас возникнут вопросы, то пишите мне в Пачку. Можете связаться как Вам удобно, либо организованно через тимлида, либо задать вопросы отдельно.

AlexeiRomanchenko committed:
fix: добавлено описание проекта в файле README.md

elGordoGato committed:
Merge branch 'develop' of https://github.com/AlexeiRomanchenko/java-filmorate into add-search

ARGalimov committed:
Исправление замечаний ревьювера в функциональностях "Отзывы" и "Лента событий"

elGordoGato committed:
refactor: add enum SearchParam

ARGalimov committed:
Merge pull request #33 from AlexeiRomanchenko/add-feed

Исправление замечаний ревьювера в функциональностях "Отзывы" и "Лента…

AlexeiRomanchenko committed:
fix: Передача параметров при логирование через placeholder - {}

elGordoGato committed:
Refactor: add enum SearchParam for findSearchedFilm method (#34)

 refactor: add enum SearchParam

AlexeiRomanchenko committed:
Merge remote-tracking branch 'origin/develop' into develop

AlexeiRomanchenko committed:
fix: Добавлена анатация @Positive. Удалены лишние поля. Исправлено логирование

AlexeiRomanchenko committed:
fix: Добавлена анатация @Positive. Удалены лишние поля. Исправлено логирование

AlexeiRomanchenko committed:
fix: Добавлена анатация @Positive. Удалены лишние поля. Исправлено логирование

elGordoGato committed:
Merge branch 'develop' of https://github.com/AlexeiRomanchenko/java-filmorate into add-search

AlexeiRomanchenko committed:
fix:

AlexeiRomanchenko committed:
fix: добавлена анотация @RequiredArgsConstructor

elGordoGato committed:
refactor: add loadAllDirectors method
fix: N+1 problem

elGordoGato committed:
Merge branch 'develop' of https://github.com/AlexeiRomanchenko/java-filmorate into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/db/DirectorDbStorage.java

elGordoGato committed:
refactor: codestyle

ARGalimov committed:
Исправление замечаний ревьювера в функциональностях "Отзывы" и "Лента событий"

ARGalimov committed:
Исправление замечаний ревьювера в функциональностях "Отзывы" и "Лента событий"

ARGalimov committed:
Merge branch 'develop' into add-feed

ARGalimov committed:
Исправление замечаний ревьювера в функциональностях "Отзывы" и "Лента событий"

ARGalimov committed:
Merge remote-tracking branch 'origin/add-feed' into add-feed

ARGalimov committed:
Исправление замечаний ревьювера в функциональностях "Отзывы" и "Лента событий"

ARGalimov committed:
Исправление замечаний ревьювера в функциональностях "Отзывы" и "Лента событий"

ARGalimov committed:
Исправление замечаний ревьювера в функциональностях "Отзывы" и "Лента событий"

ARGalimov committed:
Исправление замечаний ревьювера в функциональностях "Отзывы" и "Лента событий"

ARGalimov committed:
Исправление замечаний ревьювера в функциональностях "Отзывы" и "Лента событий"

ARGalimov committed:
Исправление замечаний ревьювера в функциональностях "Отзывы" и "Лента событий"

elGordoGato committed:
refactor: add rsMapper, findById return Optional

ARGalimov committed:
Исправление замечаний ревьювера в функциональностях "Отзывы" и "Лента событий"

ARGalimov committed:
Merge pull request #38 from AlexeiRomanchenko/add-feed

Add feed

elGordoGato committed:
Merge branch 'develop' of https://github.com/AlexeiRomanchenko/java-filmorate into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/db/EventDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/db/ReviewDbStorage.java

elGordoGato committed:
refactor: final merge

elGordoGato committed:
refactor: codestyle

elGordoGato committed:
Исправление N+1 проблемы (#35)

refactor: add enum SearchParam
add loadAllDirectors method
fix: N+1 problem
add rsMapper, findById return Optional

elGordoGato committed:
Merge branch 'develop' of https://github.com/AlexeiRomanchenko/java-filmorate into add-search

elGordoGato committed:
fix: postman errors

AlexeiRomanchenko committed:
Merge pull request #39 from AlexeiRomanchenko/add-search

Fix: Postman errors

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/DirectorController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @GetMapping("/directors/{id}")
+    public Director getDirectorById(@Positive @PathVariable int id) {
```
Чтобы аннотации валидации на подобие `@Positive` работали, надо поставить над классом аннотацию `@Validated`

ARGalimov replied:
Добавил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/MPAController.java:
```diff
@@ -28,8 +29,8 @@ public List<RatingMpa> getRatingsMpa() {
     @GetMapping("/{id}")
-    public RatingMpa getRatingMpaById(@PathVariable Integer id) {
-        log.info(LogMPA.GET_MPA_REQUEST.getMessage() + id);
+    public RatingMpa getRatingMpaById(@Positive @PathVariable Integer id) {
```
Чтобы аннотации валидации на подобие `@Positive` работали, надо поставить над классом аннотацию `@Validated`

ARGalimov replied:
Добавил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @GetMapping("/{id}")
+    public Review findById(@Positive @PathVariable Integer id) {
```
Чтобы аннотации валидации на подобие `@Positive` работали, надо поставить над классом аннотацию `@Validated`

ARGalimov replied:
Добавил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @PutMapping("/{id}/like/{userId}")
+    public void addLike(
```
Если хотите - можно сформировать отдельный контроллер для оценки - лайков и дизлайков

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -40,40 +41,82 @@ public User update(@Valid @RequestBody User user) {
     }
 
+    @DeleteMapping("/{userId}")
+    public void deleteUser(@Positive @PathVariable int userId) {
```
Чтобы аннотации валидации на подобие `@Positive` работали, надо поставить над классом аннотацию `@Validated`

ARGalimov replied:
Добавил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,26 +1,27 @@
 @RestController
 @RequestMapping("/users")
-@Slf4j
+@RequiredArgsConstructor
```
Стоит предпочесть данную аннотацию во всех контроллерах, для единства стиля проекта

ARGalimov replied:
Поправил во всех контроллерах

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -40,40 +41,82 @@ public User update(@Valid @RequestBody User user) {
         return userService.findUser(id);
     }
 
     @PutMapping(value = "/{id}/friends/{friendId}")
```
Рекомендую сформировать отдельный контроллер для работы с друзьями

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/description/LogDirector.java:
```diff
@@ -0,0 +1,38 @@
+
+    private final String message;
+
+    LogDirector(String message) {
```
Можно так же воспользоваться `@RequiredArgsConstructor`
Аналогичный совет для всех подобных мест

ARGalimov replied:
Добавил во всех подобных местах

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,83 +1,119 @@
-
-        if (userId < 0) {
-            throw new ObjectNotFoundException(LogMessagesUsers.ID_NOT_POSITIVE.getMessage());
+        if (filmStorage.removeLike(filmId, userId) == 0) {
```
Можно вернуть `null`, если лайк не был найден и удален, ну и потом использовать конструкцию `orElseThrow()`, если хотите

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,149 @@
+        validationBeforeUpdate(review);
+        Review newData = reviewStorage.update(review);
+        if (newData == null) {
+            log.warn(LogMessagesReviews.MSG_ERR_NOT_FOUND.getMessage() + review.getReviewId());
```
Здесь так же можно воспользоваться placeholder-ом, стоит выдержать единый стиль во всем приложении

ARGalimov replied:
поправил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,149 @@
+        validateId(review.getReviewId());
+    }
+
+    public void validateId(Integer id) {
```
Если метод используется в только данном классе - стоит его сделать `private`, так же если мы валидируем `id` в контроллере, нужно ли его валидировать здесь повторно?

ARGalimov replied:
Поменял на private

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    public void validateId(Integer id) {
+        if (id == null) {
```
На `null` можно проверять в модели данных за счёт аннотации `@NonNull`

ARGalimov replied:
Поправил

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/mapper/FilmMapper.java:
```diff
@@ -0,0 +1,73 @@
+@Component
+@RequiredArgsConstructor
+public class FilmMapper {
+    private final FilmDbStorage filmStorage;
```
Данное поле лишнее - в данном маппере нет обращений к БД (и не должно быть)

ARGalimov replied:
Убрал

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/mapper/FilmMapper.java:
```diff
@@ -0,0 +1,73 @@
+import java.util.Set;
+
+
+@Component
```
Данная аннотация лишняя, стоит её убрать
Все методы в этом классе делаем статическими, чтобы использовать данный класс только для преобразования

ARGalimov replied:
Убрал

KirPyt reviewed (commented):
Очень хорошо!
Указал ещё некоторые моменты, над которыми стоит поработать, ничего критичного, но в целом это поможет выдержать единый стиль проекта и уберёт лишний код

ARGalimov committed:
Исправление замечаний ревьювера в функциональностях "Отзывы" и "Лента событий"

ARGalimov committed:
Исправление замечаний ревьювера в функциональностях "Отзывы" и "Лента событий"

AlexeiRomanchenko committed:
Merge pull request #40 from AlexeiRomanchenko/add-feed

Исправление замечаний ревьювера в функциональностях "Отзывы" и "Лента…

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/DirectorDbStorage.java:
```diff
@@ -0,0 +1,174 @@
+            jdbcTemplate.update(requestUpdateDirector, director.getName(), director.getId());
+        } catch (DataIntegrityViolationException | BadSqlGrammarException ex) {
+            updateDirector(bufferDirector);
+            throw new RuntimeException(LogSQL.SQL_EXCEPTION.getMessage());
```
Стоит выбрасывать более конкретно исключение, рекомендую определять свои исключения для таких случаев

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/ReviewDbStorage.java:
```diff
@@ -0,0 +1,152 @@
+import java.util.stream.Collectors;
+
+@Component
+@Primary
```
Данная аннотация лишняя - так как других реализаций данного компонента у нас нет

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/db/mapper/FilmMapper.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.Objects;
+import java.util.Set;
+
+@RequiredArgsConstructor
```
В данном классе конструктор не требуется

KirPyt reviewed (commented):
Вы продемонстрировали очень хорошее владение навыками разработки, полученными на прошлых проектах!

Стоит внимательно ещё раз повторить работу с sql - это очень важный и нужный навык, советую присмотреться к оптимизации запросов (это в целом как рекомендация), так же важным навыком, который Вы освоили на данном групповом проекте является перекрёстное ревью - очень важный навык, так как разбираться в чужом коде нужно уметь, этого в работе разработчика достаточно) Постарайтесь позадавать друг другу вопросы и узнать, почему в том или ином месте Ваш коллега решил реализовать ту или иную логику, каковы аргументы того или иного подхода. В остальном всё хорошо, проект сложный, много тестов, но и это неплохое испытание для проверки своих навыков. Так же не забудьте ещё раз ознакомиться с конвенцией о коммитах - это нужно)

Хочу пожелать Вам удачи в дальнейших проектах и проявлять такой же ответственный подход к разработке, как и на этом проекте!
